sdcflows/transform.py

```diff
@@ -72,151 +183,189 @@ def fit(self, spatialimage):
+        ro_time : :obj:`float`
+            The total readout time in seconds.
+        xfms : :obj:`None` or :obj:`list`
+            A list of rigid-body transformations previously estimated that will
+            realign the dataset (that is, compensate for head motion) after resampling.
+        order : :obj:`int`, optional
             The order of the spline interpolation, default is 3.
             The order has to be in the range 0-5.
         mode : {'constant', 'reflect', 'nearest', 'mirror', 'wrap'}, optional
             Determines how the input image is extended when the resamplings overflows
             a border. Default is 'constant'.
         cval : float, optional
             Constant value for ``mode='constant'``. Default is 0.0.
-        prefilter: bool, optional
+        prefilter : :obj:`bool`, optional
             Determines if the image's data array is prefiltered with
             a spline filter before interpolation. The default is ``True``,
             which will create a temporary *float64* array of filtered values
             if *order > 1*. If setting this to ``False``, the output will be
             slightly blurred if *order > 1*, unless the input is prefiltered,
             i.e. it is the result of calling the spline filter on the original
             input.
+        output_dtype : :obj:`str` or :obj:`~numpy.dtype`
+            Override the output data type, instead of propagating it from the
+            moving image.
 
         Returns
         -------
-        resampled : `spatialimage` or ndarray
+        resampled : :obj:`~nibabel.spatialimages.SpatialImage`
             The data imaged after resampling to reference space.
 
         """
         from sdcflows.utils.tools import ensure_positive_cosines
 
         # Ensure the fmap has been computed
-        if isinstance(spatialimage, (str, bytes, Path)):
-            spatialimage = nb.load(spatialimage)
-
-        spatialimage, axcodes = ensure_positive_cosines(spatialimage)
+        if isinstance(moving, (str, bytes, Path)):
+            moving = nb.load(moving)
 
-        self.fit(spatialimage)
-        fmap = self.mapped.get_fdata().copy()
+        # Generate warp field (before ensuring positive cosines)
+        self.fit(moving)
 
-        # Reverse mapped if reversed blips
-        if pe_dir.endswith("-"):
-            fmap *= -1.0
+        # Make sure the data array has all cosines positive (i.e., no axes are flipped)
+        moving, axcodes = ensure_positive_cosines(moving)
 
-        # Generate warp field
         pe_axis = "ijk".index(pe_dir[0])
+        axis_flip = axcodes[pe_axis] in ("LPI")
+        pe_flip = pe_dir.endswith("-")
+
+        # Displacements are reversed if either is true (after ensuring positive cosines)
+        if axis_flip ^ pe_flip:
+            ro_time *= -1.0
+
+        # Prepare data
+        data = np.squeeze(np.asanyarray(moving.dataobj))
+        ndim = min(data.ndim, 3)
```

Comment:
I think you really just want to drop any trailing 1s after the 3rd dimension, correct? On the off chance you do have a 1 in your spatial dimensions, you don't want to invalidate your affine.

```suggestion
        # Squeeze non-spatial dimensions
        newshape = data.shape[:3] + tuple(dim for dim in data.shape[3:] if dim > 1)
        data = np.reshape(moving.dataobj, newshape)
        ndim = min(data.ndim, 3)
```

Reply:
Yes. I had some interest in making this generalizable to 2D but disregarded the idea after realizing it was really hard to infer from the image only and probably there are other decision points where 3D is assumed. Thanks for the improvement.